File: js-ts-automation/playwrite/Pages/InventoryPage.ts
import { Page, Locator, expect } from '@playwright/test';
import { URLS } from '../data/testData';

export type SortOption = 'az' | 'za' | 'lohi' | 'hilo';

/**
 * InventoryPage — product listing, sorting, and add-to-cart interactions.
 */
export class InventoryPage {
  private readonly page: Page;

  readonly sortDropdown = () => this.page.getByTestId('product-sort-container');
  readonly cartBadge = () => this.page.locator('.shopping_cart_badge');
  readonly cartLink = () => this.page.locator('.shopping_cart_link');
  readonly productItems = () => this.page.locator('.inventory_item');
  readonly productNames = () => this.page.locator('.inventory_item_name');
  readonly productPrices = () => this.page.locator('.inventory_item_price');
  readonly burgerMenuButton = () => this.page.getByRole('button', { name: 'Open Menu' });
  readonly resetAppStateLink = () => this.page.getByRole('link', { name: 'Reset App State' });
  readonly logoutLink = () => this.page.getByRole('link', { name: 'Logout' });

  constructor(page: Page) {
    this.page = page;
  }

  async goto(): Promise<void> {
    await this.page.goto(URLS.inventory);
  }

  /** Select a sort option from the dropdown */
  async sortBy(option: SortOption): Promise<void> {
    await this.sortDropdown().selectOption(option);
  }

  /** Click the Add-to-Cart button for a product by its data-test id */
  async addToCartByDataTest(dataTestId: string): Promise<void> {
    await this.page.getByTestId(dataTestId).click();
  }

  /** Add multiple products by passing an array of data-test ids */
  async addMultipleToCart(dataTestIds: string[]): Promise<void> {
    for (const id of dataTestIds) {
      await this.addToCartByDataTest(id);
    }
  }

  /** Remove an item from the inventory page via its remove button data-test id */
  async removeFromCartByDataTest(dataTestId: string): Promise<void> {
    await this.page.getByTestId(dataTestId).click();
  }

  /** Click an item name to open the product detail page */
  async openProductDetail(productName: string): Promise<void> {
    await this.page.getByText(productName, { exact: true }).click();
  }

  /** Navigate to cart */
  async goToCart(): Promise<void> {
    await this.cartLink().click();
    await expect(this.page).toHaveURL(URLS.cart);
  }

  /** Open burger menu and click Logout */
  async logout(): Promise<void> {
    await this.burgerMenuButton().click();
    await this.logoutLink().click();
    await expect(this.page).toHaveURL(URLS.base);
  }

  /** Open burger menu and reset app state (clears cart without navigating) */
  async resetAppState(): Promise<void> {
    await this.burgerMenuButton().click();
    await this.resetAppStateLink().click();
  }

  // ─── Assertions ──────────────────────────────────────────────────────────────

  async assertCartBadgeCount(expected: number): Promise<void> {
    await expect(this.cartBadge()).toBeVisible();
    await expect(this.cartBadge()).toHaveText(String(expected));
  }

  async assertCartBadgeNotVisible(): Promise<void> {
    await expect(this.cartBadge()).not.toBeVisible();
  }

  /**
   * Retrieve all prices as numbers and assert they are sorted descending.
   * Returns the parsed array so tests can do further assertions.
   */
  async assertPricesSortedDescending(): Promise<number[]> {
    const rawPrices = await this.productPrices().allTextContents();
    const prices = rawPrices.map(p => parseFloat(p.replace('$', '')));
    const isSorted = prices.every((val, i, arr) => i === 0 || arr[i - 1] >= val);
    expect(isSorted, `Prices not sorted descending: ${prices}`).toBe(true);
    return prices;
  }

  async assertPricesSortedAscending(): Promise<number[]> {
    const rawPrices = await this.productPrices().allTextContents();
    const prices = rawPrices.map(p => parseFloat(p.replace('$', '')));
    const isSorted = prices.every((val, i, arr) => i === 0 || arr[i - 1] <= val);
    expect(isSorted, `Prices not sorted ascending: ${prices}`).toBe(true);
    return prices;
  }

  async assertNamesAlphabeticallyAscending(): Promise<string[]> {
    const names = await this.productNames().allTextContents();
    const sorted = [...names].sort((a, b) => a.localeCompare(b));
    expect(names).toEqual(sorted);
    return names;
  }

  async assertNamesAlphabeticallyDescending(): Promise<string[]> {
    const names = await this.productNames().allTextContents();
    const sorted = [...names].sort((a, b) => b.localeCompare(a));
    expect(names).toEqual(sorted);
    return names;
  }

  async assertProductCount(expected: number): Promise<void> {
    await expect(this.productItems()).toHaveCount(expected);
  }
}

File: reqnroll-automation/playwright/Pages/InventoryPage.cs
using Microsoft.Playwright;
using NUnit.Framework;
using ReqnrollPlaywright.Fixtures;

namespace ReqnrollPlaywright.Pages;

/// <summary>
/// InventoryPage — product listing, sorting, and add-to-cart interactions.
/// </summary>
public class InventoryPage
{
    private readonly IPage _page;

    public InventoryPage(IPage page)
    {
        _page = page;
    }

    public ILocator SortDropdown => _page.Locator("[data-test='product-sort-container']");
    public ILocator CartBadge => _page.Locator(".shopping_cart_badge");
    public ILocator CartLink => _page.Locator(".shopping_cart_link");
    public ILocator ProductItems => _page.Locator(".inventory_item");
    public ILocator ProductNames => _page.Locator(".inventory_item_name");
    public ILocator ProductPrices => _page.Locator(".inventory_item_price");
    public ILocator BurgerMenuButton => _page.GetByRole(AriaRole.Button, new() { Name = "Open Menu" });
    public ILocator ResetAppStateLink => _page.GetByRole(AriaRole.Link, new() { Name = "Reset App State" });
    public ILocator LogoutLink => _page.GetByRole(AriaRole.Link, new() { Name = "Logout" });

    public async Task GoToAsync() => await _page.GotoAsync(Urls.Inventory);

    /// <summary>Select a sort option from the dropdown.</summary>
    public async Task SortByAsync(string option)
    {
        await SortDropdown.SelectOptionAsync(option);
    }

    /// <summary>Click the Add-to-Cart button for a product by its data-test id.</summary>
    public async Task AddToCartByDataTestAsync(string dataTestId)
    {
        await _page.Locator($"[data-test='{dataTestId}']").ClickAsync();
    }

    /// <summary>Add multiple products by passing a collection of data-test ids.</summary>
    public async Task AddMultipleToCartAsync(IEnumerable<string> dataTestIds)
    {
        foreach (var id in dataTestIds)
        {
            await AddToCartByDataTestAsync(id);
        }
    }

    /// <summary>Remove an item from the inventory page via its remove button data-test id.</summary>
    public async Task RemoveFromCartByDataTestAsync(string dataTestId)
    {
        await _page.Locator($"[data-test='{dataTestId}']").ClickAsync();
    }

    /// <summary>Click an item name to open the product detail page.</summary>
    public async Task OpenProductDetailAsync(string productName)
    {
        await _page.GetByText(productName, new() { Exact = true }).ClickAsync();
    }

    /// <summary>Navigate to cart.</summary>
    public async Task GoToCartAsync()
    {
        await CartLink.ClickAsync();
        Assert.That(_page.Url, Is.EqualTo(Urls.Cart));
    }

    /// <summary>Open burger menu and click Logout.</summary>
    public async Task LogoutAsync()
    {
        await BurgerMenuButton.ClickAsync();
        await LogoutLink.ClickAsync();
        Assert.That(_page.Url, Is.EqualTo(Urls.Base));
    }

    /// <summary>Open burger menu and reset app state (clears cart without navigating).</summary>
    public async Task ResetAppStateAsync()
    {
        await BurgerMenuButton.ClickAsync();
        await ResetAppStateLink.ClickAsync();
    }

    // --- Assertions ---------------------------------------------------------

    public async Task AssertProductCountAsync(int expected)
    {
        Assert.That(await ProductItems.CountAsync(), Is.EqualTo(expected));
    }

    public async Task AssertCartBadgeCountAsync(int expected)
    {
        Assert.That(await CartBadge.IsVisibleAsync(), Is.True);
        Assert.That(await CartBadge.TextContentAsync(), Is.EqualTo(expected.ToString()));
    }

    public async Task AssertCartBadgeNotVisibleAsync()
    {
        Assert.That(await CartBadge.IsVisibleAsync(), Is.False);
    }

    /// <summary>
    /// Retrieve all prices as numbers and assert they are sorted descending.
    /// Returns the parsed list so tests can do further assertions.
    /// </summary>
    public async Task<List<double>> AssertPricesSortedDescendingAsync()
    {
        var prices = await GetPricesAsync();
        var isSorted = prices.Zip(prices.Skip(1), (a, b) => a >= b).All(ordered => ordered);
        Assert.That(isSorted, Is.True, $"Prices not sorted descending: {string.Join(", ", prices)}");
        return prices;
    }

    public async Task<List<double>> AssertPricesSortedAscendingAsync()
    {
        var prices = await GetPricesAsync();
        var isSorted = prices.Zip(prices.Skip(1), (a, b) => a <= b).All(ordered => ordered);
        Assert.That(isSorted, Is.True, $"Prices not sorted ascending: {string.Join(", ", prices)}");
        return prices;
    }

    public async Task<List<string>> AssertNamesAlphabeticallyAscendingAsync()
    {
        var names = (await ProductNames.AllTextContentsAsync()).ToList();
        var sorted = names.OrderBy(n => n, StringComparer.Ordinal).ToList();
        Assert.That(names, Is.EqualTo(sorted));
        return names;
    }

    public async Task<List<string>> AssertNamesAlphabeticallyDescendingAsync()
    {
        var names = (await ProductNames.AllTextContentsAsync()).ToList();
        var sorted = names.OrderByDescending(n => n, StringComparer.Ordinal).ToList();
        Assert.That(names, Is.EqualTo(sorted));
        return names;
    }

    private async Task<List<double>> GetPricesAsync()
    {
        var raw = await ProductPrices.AllTextContentsAsync();
        return raw
            .Select(p => double.Parse(p.Replace("$", string.Empty).Trim(),
                System.Globalization.CultureInfo.InvariantCulture))
            .ToList();
    }
}
